singularity-eos/eos/eos_jwl.hpp

```diff
@@ -96,22 +106,21 @@ class JWL : public EosBase<JWL> {
   static std::string EosPyType() { return EosType(); }
 
  private:
-  Real _A, _B, _R1, _R2, _w, _rho0, _Cv;
+  Real _A, _B, _R1, _R2, _w, _rho0, _Cv, _c1, _c2;
   PORTABLE_INLINE_FUNCTION Real ReferenceEnergy(const Real rho) const;
   PORTABLE_INLINE_FUNCTION Real ReferencePressure(const Real rho) const;
   // static constexpr const char _eos_type[] = "JWL";
   static constexpr const unsigned long _preferred_input =
       thermalqs::density | thermalqs::specific_internal_energy;
 };
 
-PORTABLE_INLINE_FUNCTION Real JWL::ReferencePressure(const Real rho) const {
-  const Real x = _rho0 / rho;
-  return _A * std::exp(-_R1 * x) + _B * std::exp(-_R2 * x);
+PORTABLE_FORCEINLINE_FUNCTION Real JWL::ReferencePressure(const Real rho) const {
+  const Real x = robust::ratio(_rho0, rho);
+  return _A * robust::safe_arg_exp(-_R1 * x) + _B * robust::safe_arg_exp(-_R2 * x);
 }
-PORTABLE_INLINE_FUNCTION Real JWL::ReferenceEnergy(const Real rho) const {
-  const Real x = _rho0 / rho;
-  return _A / (_rho0 * _R1) * std::exp(-_R1 * x) +
-         _B / (_rho0 * _R2) * std::exp(-_R2 * x);
+PORTABLE_FORCEINLINE_FUNCTION Real JWL::ReferenceEnergy(const Real rho) const {
+  const Real x = robust::ratio(_rho0, rho);
+  return _c1 * robust::safe_arg_exp(-_R1 * x) + _c2 * robust::safe_arg_exp(-_R2 * x);
```

Comment:
This equation (and the one for pressure) could still overflow if `robust::safe_arg_exp` returns `std::numeric_limits<T>::max()`. I see three ways to solve this:
1) Split the result and max at the numeric limit at each stage
2) Have the safe exponent return infinity and then add a ternary expression to return the max if infinity is reached
3) Test the arguments before putting them into the exponent and just return the maximum value if they're going to overflow.

Reply:
oof this opens up such a can of worms for every mathematical expression in the code. In a simple example, making use of infinity could be a way around FPE's. And all of this to *not* error out on a code that will pass garbage data with FPE trapping enabled. I think this is targeting a discussion much larger than the intended scope of this PR.

Reply:
why not just have the exp return `std::numeric_limits<T>::max() / 10` as the maximum allowed output? Doesn't solve the problem necessarily, but less likely to overflow/underflow at least.

Reply:
I think returning infinity fixes it as well.

Reply:
Won't infinity trigger a floating point exception? If it doesn't, I agree---that's the better solution. It's what infinity is for.

Reply:
My experience is more with the numpy implementation, but since the c++ standard and the numpy documentation both cite IEEE 754, I _think_ it shouldn't trigger FPEs. I'm not 100% sure about this though.

On this point, I remember trying to get compiler explore to let me trigger FPEs and I remember it being hard. Does anybody have a working example to do this?

Reply:
> I think that is asking too much @jhp-lanl

I agree 😄 . Which is why I _think_ using infinity is the wise choice here. I _think_ infinity fails the IEEE normal check, but doesn't trigger FPEs, which I think is desirable here.